File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        i = 0
        j = len(height) -1
        res = 0
        lmax = 0
        rmax = 0
        while i<=j:
            if height[i] < height[j]:
                if height[i] >= lmax:
                    lmax = height[i]
                else:
                    res += (lmax - height[i])
                i+=1
            else:
                if height[j] >= rmax:
                    rmax = height[j]
                else:
                    res += (rmax - height[j])
                j -=1
        return res
            
            
        

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int res = 0;
        int len = height.length;
        int[] rmax = new int[len];
        int[] lmax = new int[len];
        int lm = height[0];
        for(int i = 1; i<len; i++) {
            lmax[i] = lm;
            lm = Math.max(lm, height[i]);
        }
        int rm = height[len-1];
        for(int i = len-2; i>=0; i--) {
            rmax[i] = rm;
            rm = Math.max(rm, height[i]);
        }
        for(int i = 0; i<height.length; i++) {
            int min = Math.min(lmax[i], rmax[i]);
            int water = min - height[i];
            if(water > 0) {
                res += water;
            }
        }
        return res;
    }
}